File: GrooveWorkAndroid/app/src/main/java/com/kodass/groovework/ui/calendars/CalendarFields.kt
package com.kodass.groovework.ui.calendars

import com.kodass.groovework.data.dto.CalendarDto
import com.kodass.groovework.data.dto.CalendarEntryDto
import com.kodass.groovework.data.dto.CalendarFieldDto
import com.kodass.groovework.ui.registries.textValue
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.contentOrNull
import java.time.Instant
import java.time.LocalDate
import java.time.OffsetDateTime
import java.time.ZoneId
import java.time.format.DateTimeFormatter

// Утилиты раздела «Календари» — зеркало front/src/utils/calendarFields.js.

// isFieldVisible — показывать ли поле при текущих значениях. Правило условной
// видимости: поле-источник visibleFieldId должно иметь значение, равное
// visibleValue. Для checkbox-источника visibleValue == "true", для select —
// выбранный вариант. Без условия — поле видно всегда.
fun isFieldVisible(field: CalendarFieldDto, data: Map<String, JsonElement>): Boolean {
    val src = field.visibleFieldId ?: return true
    val target = field.visibleValue ?: ""
    return when (val v = data[src.toString()]) {
        is JsonArray -> v.any { (it as? JsonPrimitive)?.contentOrNull == target }
        is JsonPrimitive -> {
            val b = v.booleanOrNull
            if (b != null) b.toString() == target else (v.contentOrNull ?: "") == target
        }
        else -> "" == target
    }
}

// entryTitle — заголовок записи для плитки/списка: первое поле «показывать в
// таблице» (иначе первое вообще); пусто → запасной текст.
fun entryTitle(calendar: CalendarDto?, entry: CalendarEntryDto, fallback: String = "Запись"): String {
    val fields = calendar?.fields ?: emptyList()
    val pick = fields.firstOrNull { it.showInTable } ?: fields.firstOrNull()
    if (pick != null) {
        val v = textValue(pick.asRegistryField(), entry.data[pick.key])
        if (v.isNotBlank()) return v
    }
    return fallback
}

// entrySubtitle — второе «таблничное» поле (для режима «День»).
fun entrySubtitle(calendar: CalendarDto?, entry: CalendarEntryDto): String {
    val fields = (calendar?.fields ?: emptyList()).filter { it.showInTable }
    val f = fields.getOrNull(1) ?: return ""
    return textValue(f.asRegistryField(), entry.data[f.key])
}

fun parseEventInstant(iso: String?): Instant? {
    if (iso.isNullOrBlank()) return null
    return runCatching { Instant.parse(iso) }
        .recoverCatching { OffsetDateTime.parse(iso).toInstant() }
        .getOrNull()
}

// Локальная дата записи (для группировки по дням).
fun entryLocalDate(entry: CalendarEntryDto): LocalDate? =
    parseEventInstant(entry.eventAt)?.atZone(ZoneId.systemDefault())?.toLocalDate()

// Время записи без секунд.
fun hhmm(iso: String?): String {
    val inst = parseEventInstant(iso) ?: return ""
    return inst.atZone(ZoneId.systemDefault()).format(TIME_HM)
}

private val TIME_HM = DateTimeFormatter.ofPattern("HH:mm")

File: front/src/utils/calendarFields.js
// Поля карточки записи календаря используют тот же набор типов и хелперы, что и
// реестры (см. utils/registryFields.js) — переиспользуем их, не дублируя.
// Здесь — только специфика календаря: условная видимость полей и заголовок
// записи для плитки.
export {
  FIELD_TYPES,
  FIELD_DATETIME,
  fieldMeta,
  fieldLabel,
  fieldIcon,
  defaultConfig,
  formatDateTime,
  textValue,
  isSearchable,
  isSortable,
  isExportable,
} from './registryFields.js'

import { textValue } from './registryFields.js'

// isFieldVisible — показывать ли поле в карточке при текущих значениях.
// Правило условной видимости: поле visible_field_id должно иметь значение,
// равное visible_value. Для checkbox-источника visible_value == 'true'
// («когда галочка отмечена»), для select — выбранный вариант. Без условия —
// поле видно всегда.
export function isFieldVisible(field, data) {
  const src = field?.visible_field_id
  if (src == null) return true
  const v = data?.[String(src)]
  const target = field.visible_value ?? ''
  if (Array.isArray(v)) return v.some((x) => String(x) === String(target))
  if (typeof v === 'boolean') return String(v) === String(target)
  return String(v ?? '') === String(target)
}

// canBeCondition — может ли поле быть источником условия (checkbox/select).
export function canBeCondition(type) {
  return type === 'checkbox' || type === 'select'
}

// entryTitle — заголовок записи для плитки/списка: первое поле, помеченное
// «показывать в таблице», иначе первое поле вообще; пусто → запасной текст.
export function entryTitle(calendar, entry, fallback = 'Запись') {
  const fields = calendar?.fields || []
  const pick = fields.find((f) => f.show_in_table) || fields[0]
  if (pick) {
    const v = textValue(pick, entry?.data?.[String(pick.id)])
    if (v) return v
  }
  return fallback
}

// cardFields — поля для тела карточки события (виды «День»/«Неделя» и диалог
// дня): помеченные show_in_card и видимые при текущих значениях записи.
// Возвращает [{ field, value }] с непустым текстовым значением.
export function cardFields(calendar, entry) {
  const fields = calendar?.fields || []
  const data = entry?.data || {}
  // Заголовок карточки рисуется отдельно (entryTitle) — не дублируем это поле.
  const titleField = fields.find((f) => f.show_in_table) || fields[0]
  const out = []
  for (const f of fields) {
    if (f.show_in_card === false) continue
    if (titleField && f.id === titleField.id) continue
    if (!isFieldVisible(f, data)) continue
    const value = textValue(f, data[String(f.id)])
    if (value) out.push({ field: f, value })
  }
  return out
}

// hhmm — время записи без секунд (для плиток и режима «День»).
export function hhmm(value) {
  if (!value) return ''
  const d = new Date(value)
  if (isNaN(d.getTime())) return ''
  const pad = (n) => String(n).padStart(2, '0')
  return `${pad(d.getHours())}:${pad(d.getMinutes())}`
}
